cpp/src/lance/io/filter.cc

```diff
@@ -0,0 +1,63 @@
+//  Copyright 2022 Lance Authors
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+#include "lance/io/filter.h"
+
+#include <arrow/array.h>
+#include <arrow/compute/api.h>
+#include <arrow/record_batch.h>
+#include <arrow/result.h>
+
+namespace lance::io {
+
+Filter::Filter(std::shared_ptr<lance::format::Schema> schema,
+               const ::arrow::compute::Expression& filter)
+    : schema_(schema), filter_(filter) {}
+
+::arrow::Result<std::unique_ptr<Filter>> Filter::Make(const lance::format::Schema& schema,
+                                                      const ::arrow::compute::Expression& filter) {
+  if (!::arrow::compute::ExpressionHasFieldRefs(filter)) {
+    /// All scalar?
+    return nullptr;
+  }
+  auto field_refs = ::arrow::compute::FieldsInExpression(filter);
+  std::vector<std::string> columns;
+  for (auto& ref : field_refs) {
+    columns.emplace_back(std::string(*ref.name()));
+  }
+  ARROW_ASSIGN_OR_RAISE(auto filter_schema, schema.Project(columns));
+  return std::unique_ptr<Filter>(new Filter(filter_schema, filter));
```

Comment:
I forgot, how does this get wrapped into Result?

Reply:
so `Result<T>` has two constructors, one from `Status`, one from `T`. 
Here T = `std::unique_ptr<Filter>`, so it `std::move`-ed the filter into `Result`.